tracing: panic with /debug/requests on

```
panic: runtime error: index out of range

goroutine 18489024 [running]:
github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).addEvent(0xc434fb81a0, 0x207a080, 0xc446acb440, 0xc44cd00000)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:772 +0x7c0
github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace.(*trace).LazyPrintf(0xc434fb81a0, 0x2318ac5, 0x2, 0xc44cd02520, 0x1, 0x1)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:801 +0xa5
github.com/cockroachdb/cockroach/pkg/util/tracing.(*span).LogFields(0xc454ee56c0, 0xc42a30b400, 0x1, 0x1)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:353 +0x9fe
github.com/cockroachdb/cockroach/pkg/util/log.eventInternal(0x34e3100, 0xc454a656b0, 0xffffffffffff0100, 0x2335663, 0x15, 0x0, 0x0, 0x0)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:131 +0x26f
github.com/cockroachdb/cockroach/pkg/util/log.VEvent(0x34e3100, 0xc454a656b0, 0xc400000001, 0x2335663, 0x15)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:190 +0x15b
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*aggregator).Next(0xc436c11500, 0x0, 0x0, 0x0, 0xc462349180)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:270 +0x415
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x34e3040, 0xc46f209b40, 0x34e3c00, 0xc436c11500, 0x34cda80, 0xc4362d4fc0)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:159 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*aggregator).Run(0xc436c11500, 0xc42bb3b040)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:191 +0x6b
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:466 +0x3c5
```

The panic is in this stanza:

```
		// Discard the middle events.
		di := int((tr.maxEvents - 1) / 2)
		if d, ok := tr.events[di].What.(*discarded); ok {
			(*d)++
```

No idea what's going on here. Triggered when looking at `/debug/requests` on a cluster with TPCC running, with `SET CLUSTER SETTING trace.debug.enable = 'true'`.

--- a/pkg/sql/distsqlrun/aggregator.go
+++ b/pkg/sql/distsqlrun/aggregator.go
@@ -203,6 +203,7 @@ func (ag *aggregator) close() {
 				aggFunc.Close(ag.ctx)
 			}
 		}
+		ag.accumulating = false
 	}
 	ag.internalClose()
 }

--- a/pkg/sql/distsqlrun/algebraic_set_op.go
+++ b/pkg/sql/distsqlrun/algebraic_set_op.go
@@ -132,10 +132,13 @@ func (e *algebraicSetOp) producerMeta(err error) *ProducerMetadata {
 }
 
 func (e *algebraicSetOp) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
-	var row sqlbase.EncDatumRow
-	var meta *ProducerMetadata
+	if e.closed {
+		return nil, e.producerMeta(nil /* err */)
+	}
 
 	for {
+		var row sqlbase.EncDatumRow
+		var meta *ProducerMetadata
 		switch e.opType {
 		case AlgebraicSetOpSpec_Except_all:
 			row, meta = e.nextExceptAll()
@@ -198,7 +201,6 @@ func (e *algebraicSetOp) nextExceptAll() (sqlbase.EncDatumRow, *ProducerMetadata
 		for i := range e.allCols {
 			e.allCols[i] = uint32(i)
 		}
-
 	}
 
 	// The loop below forms a restartable state machine that iterates over a

--- a/pkg/sql/distsqlrun/distinct.go
+++ b/pkg/sql/distsqlrun/distinct.go
@@ -152,6 +152,10 @@ func (d *distinct) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
 		d.evalCtx = d.flowCtx.NewEvalCtx()
 	}
 
+	if d.closed {
+		return nil, d.producerMeta(nil /* err */)
+	}
+
 	for {
 		row, meta := d.input.Next()
 		if d.closed || meta != nil {

--- a/pkg/sql/distsqlrun/mergejoiner.go
+++ b/pkg/sql/distsqlrun/mergejoiner.go
@@ -138,6 +138,10 @@ func (m *mergeJoiner) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
 		log.VEventf(m.ctx, 2, "starting merge joiner run")
 	}
 
+	if m.closed {
+		return nil, m.producerMeta(nil /* err */)
+	}
+
 	for {
 		row, meta := m.nextRow()
 		if m.closed || meta != nil {

--- a/pkg/sql/distsqlrun/processors.go
+++ b/pkg/sql/distsqlrun/processors.go
@@ -383,6 +383,7 @@ func (pb *processorBase) init(
 	output RowReceiver,
 ) error {
 	pb.flowCtx = flowCtx
+	pb.ctx = pb.flowCtx.Ctx
 	if evalCtx == nil {
 		evalCtx = flowCtx.NewEvalCtx()
 	}
@@ -420,8 +421,12 @@ func (pb *processorBase) internalClose() bool {
 	closing := !pb.closed
 	if closing {
 		pb.closed = true
+		pb.started = true // a closed processor has definitely started
 		tracing.FinishSpan(pb.span)
 		pb.span = nil
+		// Reset the context so that any incidental uses after this point do not
+		// access the finished span.
+		pb.ctx = pb.flowCtx.Ctx
 	}
 	// This prevents Next() from returning more rows.
 	pb.out.consumerClosed()

--- a/pkg/sql/distsqlrun/processors_test.go
+++ b/pkg/sql/distsqlrun/processors_test.go
@@ -391,3 +391,78 @@ func TestAggregatorSpecAggregationEquals(t *testing.T) {
 		}
 	}
 }
+
+func TestProcessorBaseContext(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	st := cluster.MakeTestingClusterSettings()
+
+	runTest := func(t *testing.T, f func(noop *noopProcessor)) {
+		flowCtx := &FlowCtx{
+			Ctx:      ctx,
+			Settings: st,
+			EvalCtx:  tree.MakeTestingEvalContext(st),
+		}
+		defer flowCtx.EvalCtx.Stop(ctx)
+
+		input := NewRepeatableRowSource(oneIntCol, makeIntRows(10, 1))
+		noop, err := newNoopProcessor(flowCtx, input, &PostProcessSpec{}, &RowDisposer{})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// The context should be valid before Next is called in case ConsumerDone
+		// or ConsumerClosed are called without calling Next.j
+		if noop.ctx != ctx {
+			t.Fatalf("processorBase.ctx not initialized")
+		}
+		f(noop)
+		// The context should be reset after ConsumerClosed is called so that any
+		// subsequent logging calls will not operate on closed spans.
+		if noop.ctx != ctx {
+			t.Fatalf("processorBase.ctx not reset on close")
+		}
+	}
+
+	t.Run("next-close", func(t *testing.T) {
+		runTest(t, func(noop *noopProcessor) {
+			// The normal case: a call to Next followed by the processor being closed.
+			noop.Next()
+			noop.ConsumerClosed()
+		})
+	})
+
+	t.Run("close-without-next", func(t *testing.T) {
+		runTest(t, func(noop *noopProcessor) {
+			// A processor can be closed without Next ever being called.
+			noop.ConsumerClosed()
+		})
+	})
+
+	t.Run("close-next", func(t *testing.T) {
+		runTest(t, func(noop *noopProcessor) {
+			// After the processor is closed, it can't be opened via a call to Next.
+			noop.ConsumerClosed()
+			noop.Next()
+		})
+	})
+
+	t.Run("next-close-next", func(t *testing.T) {
+		runTest(t, func(noop *noopProcessor) {
+			// A spurious call to Next after the processor is closed.
+			noop.Next()
+			noop.ConsumerClosed()
+			noop.Next()
+		})
+	})
+
+	t.Run("next-close-close", func(t *testing.T) {
+		runTest(t, func(noop *noopProcessor) {
+			// Close should be idempotent.
+			noop.Next()
+			noop.ConsumerClosed()
+			noop.ConsumerClosed()
+		})
+	})
+}

--- a/pkg/sql/distsqlrun/sorter.go
+++ b/pkg/sql/distsqlrun/sorter.go
@@ -161,13 +161,13 @@ func newSortAllProcessor(s *sorterBase) Processor {
 }
 
 func (s *sortAllProcessor) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
-	if s.closed {
-		return nil, nil
-	}
 	if s.maybeStart("sortAllProcessor", "sortAllProcessor") {
 		if err := s.fill(); err != nil {
 			return nil, s.producerMeta(err)
 		}
+	}
+	if s.closed {
+		return nil, s.producerMeta(nil /* err */)
 	}
 
 	for {
@@ -445,9 +445,13 @@ func (s *sortTopKProcessor) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
 		s.rows.Sort(ctx)
 	}
 
+	if s.closed {
+		return nil, s.producerMeta(nil /* err */)
+	}
 	if s.rows.Len() == 0 {
 		return s.trailingMetadata()
 	}
+
 	for {
 		row := s.rows.EncRow(0)
 		s.rows.PopFirst()
@@ -619,10 +623,13 @@ func (s *sortChunksProcessor) fill() (*ProducerMetadata, error) {
 
 func (s *sortChunksProcessor) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
 	s.maybeStart("sortChunks", "SortChunks")
+	if s.closed {
+		return nil, s.producerMeta(nil /* err */)
+	}
+
 	for {
 		// If we don't have an active chunk, clear and refill it.
 		if s.rows.Len() == 0 {
-
 			s.rows.Clear(s.rows.evalCtx.Ctx())
 			meta, err := s.fill()
 			if err != nil {

--- a/pkg/sql/distsqlrun/values.go
+++ b/pkg/sql/distsqlrun/values.go
@@ -101,6 +101,9 @@ func (v *valuesProcessor) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
 
 		v.rowBuf = make(sqlbase.EncDatumRow, len(v.columns))
 	}
+	if v.closed {
+		return nil, v.producerMeta(nil /* err */)
+	}
 
 	for {
 		row, meta, err := v.sd.GetRow(v.rowBuf)

